### ISSUE TITLE
test(models): audit the families upstream guards with an f32 attention widen

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1829]
## Problem / Background

lablup/mlxcel#1710 ported the upstream f32 attention guard into `src/models/phi.rs` and `src/models/stablelm.rs` (commit `3576d734`) after `phi-2-4bit` went NaN at layer 29 of 32 in f16 and every later token decoded as `!`. The defect had survived because lablup/mlxcel#1709's `gelu_approx` was widening the residual stream by accident and keeping the scores in range. That issue's description asked whether the other families upstream guards this way carry the guard in their mlxcel ports, and closed without an answer. This is that audit.

The family list in #1710 is a starting point, not the answer: a code search of the upstream trees finds a family it omits, one entry that is conditional on a config branch, and one that matches the search string without being this guard.

## Current Behavior

Upstream, searched for `queries.astype(mx.float32)` and `keys.astype(mx.float32)` on 2026-09-12. ml-explore/mlx-lm has exactly three: [`phi.py`](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/models/phi.py) (queries), [`phixtral.py`](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/models/phixtral.py) (queries), [`stablelm.py`](https://github.com/ml-explore/mlx-lm/blob/main/mlx_lm/models/stablelm.py) (queries and keys). Blaizzy/mlx-vlm has seven hits, of which six are this guard: [`phi/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/phi/language.py) (queries), [`phixtral/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/phixtral/language.py) (queries), [`stablelm/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/stablelm/language.py) (queries and keys, a family #1710 did not list), [`paligemma/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/paligemma/language.py) (queries and keys, but only in the `else` branch taken when `model_type != "gemma"`; the `"gemma"` branch calls plain `scaled_dot_product_attention` with no widen), [`minimax_m3_vl/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/minimax_m3_vl/language.py) (`idx_queries` and `idx_keys` in the block-sparse indexer only; `MiniMaxAttention` itself calls SDPA unwidened), [`nemotron_labs_diffusion/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/nemotron_labs_diffusion/language.py) (queries and keys into a manual matmul). The seventh, [`muse_glimmer/language.py`](https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/muse_glimmer/language.py), is not this guard: it multiplies by `qk_scale_factor` in f32 and casts straight back to `queries.dtype` before attention.

mlxcel at `1206f863`, for the same families:

| Family | mlxcel site | State |
|---|---|---|
| phi | `src/models/phi.rs:199` | queries widened, matches upstream |
| stablelm | `src/models/stablelm.rs:197-198` | queries and keys widened, matches upstream |
| phixtral | `src/models/phixtral.rs:497-499` | queries, keys **and** values widened; upstream widens the queries alone |
| paligemma | no model file; text attention is `src/models/gemma.rs:181-190` (PaliGemma 1) and `src/models/gemma2.rs:184-208` (PaliGemma 2) | neither widens |
| minimax_m3 indexer | `src/models/minimax_m3_indexer.rs:219` | `matmul(q, k_t)` at input dtype, no widen |
| nemotron_labs_diffusion | absent | family is not ported at all |

Two mlxcel sites that look like this guard and are not: `src/models/minimax.rs:255` is MiniMax-M2's MoE router casting gate logits, a different family from `minimax_m3_vl`; the three f32 casts in `src/models/minimax_m3_indexer.rs` (353, 440, 472) are on already-computed `token_scores` and on index arithmetic, downstream of the unwidened matmul.

## Proposed Solution

Produce a per-family verdict and land whatever the verdict requires, recorded the way lablup/mlxcel#1711's audit was in `docs/f32-promotion-audit.md`: a new `docs/f32-attention-widen-audit.md` stating, for each of the six upstream sites above, three answers.

1. **Does upstream widen there, and under what condition?** Cite the file by GitHub URL. `AGENTS.md:161-168` forbids a committed `references/...` or bare `mlx_vlm/...` path.
2. **Does the mlxcel port widen?** Cite `path:line`.
3. **If it does not, is the family reachable at a dtype where the scores can overflow?** f16's ceiling is 65504 and that is the failure mode #1710 hit. bf16 carries f32's exponent range, so on a bf16 checkpoint the same widen buys precision, not range, and the verdict must say which of the two it is rather than treating them as one case.

Expected work per family, to be confirmed or overturned by the audit rather than assumed:

- **phixtral.** `src/models/phixtral.rs:497-499` widens two operands upstream leaves alone. This is the exact over-widen `3576d734` removed from `phi.rs` and `stablelm.rs`, on the reasoning that MLX promotes the score matmul to match its wider operand so widening `q` suffices for `q @ k^T`, and values only ever multiply probabilities in `[0, 1]` and cannot overflow. Removing the `cache_k` and `cache_v` casts should be output-identical and should recover decode throughput on `phixtral-4x2_8-4bit`. Measure it; do not land it on the argument alone.
- **paligemma.** PaliGemma 1 (`text_config.model_type == "gemma"`) already matches upstream, which takes the unwidened SDPA branch there. PaliGemma 2 routes to `src/models/gemma2.rs`, which is the branch upstream widens. Gemma 2 applies `attn_logit_softcapping`, so the post-softcap scores are bounded by the cap; the open question is whether the pre-softcap `q @ k^T` can leave f16 range before the tanh, on the catalogued checkpoint `paligemma2-3b-ft-docci-448-6bit` (`docs/model-catalog.tsv:141`). A widen here also lands on every plain Gemma 2 text checkpoint that shares the module, so its throughput cost has to be measured on both paths.
- **minimax_m3.** `token_scores` at `src/models/minimax_m3_indexer.rs:219` is the site upstream widens. It is reached from `src/models/minimax_m3_layers.rs:159-165` only when `should_apply_sparse(kv_len)` is true, that is `kv_len > 2 * topk_blocks * block_size`, so it is a long-context path. The scores drive a discrete top-k block selection, which makes a wrong verdict here silent in a different way from #1710: no NaN, just different blocks selected. State whether the M3 checkpoints ship f16 or bf16 before deciding.
- **nemotron_labs_diffusion.** Not ported. mlxcel's `nemotron*` files are a different family and its diffusion LMs are `DiffusionGemma` and `Llada2Moe` (`src/models/mod.rs:425-426`). The verdict is "no port, nothing to do", and the value of recording it is that a future port starts from a written answer.

## Scope

**In scope:** `docs/f32-attention-widen-audit.md` (new); `src/models/phixtral.rs`, `src/models/gemma2.rs`, `src/models/minimax_m3_indexer.rs` if and only if the audit finds a real gap; a test per changed family.

**Out of scope:** `src/models/phi.rs` and `src/models/stablelm.rs`, settled by `3576d734`; the general f32-promotion class covered by `docs/f32-promotion-audit.md` and lablup/mlxcel#1711; benchmark table refreshes beyond the rows a landed change moves.

## Implementation Notes

- **Reuse.** Follow `docs/f32-promotion-audit.md` for shape and for tone: its useful half is the negative one, naming why each candidate is not a defect. A family that legitimately needs no widen is a result, not a gap.
- **Reuse.** `mlxcel_core::astype` with `mlxcel_core::dtype::FLOAT32` and the surrounding `array_dtype` capture and cast-back already exist at `src/models/phi.rs:198-211`. Copy that, including the comment recording why the cast is load-bearing, rather than inventing a second idiom.
- **Constraint.** The guard costs throughput and the project has accepted that where it is real: on M1 Ultra `phi-2-4bit` reads 128.76 tok/s with it against 164.10 without, and the faster number was a broken model (#1710). Any widen this audit adds must be reported with a before/after decode number on the affected checkpoint, and any widen it removes with the same.
- **Edge case.** bf16 checkpoints. Do not add an overflow-motivated widen to a family that only ever runs bf16 without saying what it buys there.
- **Edge case.** A family whose checkpoint is absent from the development machine (MiniMax-M3 is called out as exceeding it in `docs/supported-models.md:41`). Record that as "unverified, checkpoint absent" rather than as a pass. `tests/common/mod.rs:18-34` documents how the opposite mistake, a skip reported as `ok`, left an earlier NaN guard covering nothing on either machine.
- **Error handling.** None of this is a runtime error path: a wrong verdict surfaces as NaN output or as a throughput loss, never as an `Err`.

## Acceptance Criteria

- [ ] `docs/f32-attention-widen-audit.md` exists and carries a verdict for each of the six upstream guard sites, each citing the upstream file by `https://github.com/...` URL and the mlxcel counterpart by `path:line`.
- [ ] Each verdict answers all three questions: does upstream widen, does mlxcel widen, and if not, is the family reachable at a dtype whose range the scores can leave, naming f16 or bf16 explicitly.
- [ ] The doc records that `nemotron_labs_diffusion` has no mlxcel port, and that `muse_glimmer` matches the search string without being this guard, so a later re-run of the search does not re-open either.
- [ ] For every family where a widen is added or removed, a real-checkpoint run is reported, not only a unit test: #1710's defect passed every unit test in the tree for months, and the checkpoints for phi, stablelm and phixtral are present under `models/mlx/`.
- [ ] For every family where a widen is added or removed, a before/after decode tok/s on the affected checkpoint is recorded in the doc.
- [ ] Any change lands in the real forward path of the family's model file, not behind a flag or in a helper nothing calls.
- [ ] `phixtral-4x2_8-4bit` generation is byte-identical across the phixtral change, checked with `MLXCEL_PRINT_TOKEN_IDS=1` on both sides.

## Verification

```bash
cargo fmt --all -- --check
cargo clippy --workspace --all-targets --features metal,accelerate -- -D warnings
cargo test --workspace --profile test-fast --features metal,accelerate

# Real-checkpoint output check per touched family. MLXCEL_REQUIRE_MODELS=1 turns a
# missing checkpoint into a failure instead of a silent skip in the test paths.
MLXCEL_PRINT_TOKEN_IDS=1 cargo run --release --features metal,accelerate -- \
    generate -m models/mlx/phixtral-4x2_8-4bit -p "Paris is" -n 32
MLXCEL_PRINT_TOKEN_IDS=1 cargo run --release --features metal,accelerate -- \
    generate -m models/mlx/phi-2-hf-4bit-mlx -p "Paris is" -n 32
```

A pass is fluent text with no `!` run and no repeated degenerate token on every checkpoint touched, identical token ids across a change asserted to be output-neutral, and a decode number for each moved row. Upstream is a moving target, so state the date the search was re-run.

## Technical Considerations

The search behind the table above is reproducible: GitHub code search for `"queries.astype(mx.float32)"` and `"keys.astype(mx.float32)"` scoped to `repo:ml-explore/mlx-lm` and `repo:Blaizzy/mlx-vlm`. Committing it as a script beside `scripts/audit_f32_promotion.py` would let the next MLX pin bump re-run it, but a script is not required here; the written verdicts are.

Related: lablup/mlxcel#1710 (the fix this follows), lablup/mlxcel#1709 (the masking defect), lablup/mlxcel#1711 and `docs/f32-promotion-audit.md` (the adjacent f32 audit and its format).

**Comment by inureyes** [comment]
Ordering note: this audit should follow lablup/mlxcel#1830 rather than run beside it.

#1830 builds `max_abs_attention_score`, which turns the question this issue has to answer for every family from a judgement into a measurement: how much headroom the scores have against the f16 ceiling before anything is changed.

The case that makes the order matter is `phixtral`. `src/models/phixtral.rs:496-499` widens `q`, `cache_k` and `cache_v`, where upstream widens the queries alone, so it carries exactly the over-widen that `3576d734` removed from phi and stablelm. Narrowing it would buy throughput back, and phi is a precedent that narrowing can be safe. But phi was a family with no headroom at all, which is why the guard was needed there, and phixtral's headroom is simply unknown. Measuring it first is the difference between a decision and a guess.

The same applies in the other direction to any family this audit finds unguarded: the probe says whether the family is reachable in f16 with scores that can overflow, which is the question, rather than whether upstream happens to widen there.

An M5 Max host is available for the cross-host half of any family that ends up changed, with the measurement discipline recorded in lablup/mlxcel#1830's comment.
